server-spi-private/src/main/java/org/keycloak/utils/LockObjectsForModification.java

```diff
@@ -38,21 +42,36 @@ public class LockObjectsForModification {
 
     private static final String ATTRIBUTE = LockObjectsForModification.class.getCanonicalName();
 
-    public static LockObjectsForModification.Enabled enable(KeycloakSession session) {
-        return new Enabled(session);
+    public static boolean isEnabled(KeycloakSession session, Class<?> model) {
+        Set<Class<?>> lockedModels = getAttribute(session);
+        return lockedModels != null && lockedModels.contains(model);
+    }
+
+    private static Set<Class<?>> getAttribute(KeycloakSession session) {
+        //noinspection unchecked
+        return (Set<Class<?>>) session.getAttribute(ATTRIBUTE);
+    }
+
+    private static Set<Class<?>> getOrCreateAttribute(KeycloakSession session) {
+        Set<Class<?>> attribute = getAttribute(session);
+        if (attribute == null) {
+            attribute = new HashSet<>();
+            session.setAttribute(ATTRIBUTE, attribute);
+        }
+        return attribute;
     }
 
-    public static boolean isEnabled(KeycloakSession session) {
-        return session.getAttribute(ATTRIBUTE) != null;
+    public static <V> V lockUserSessionsForModification(KeycloakSession session, CallableWithoutThrowingAnException<V> callable) {
+        return lockObjectsForModification(session, UserSessionModel.class, callable);
     }
 
-    public static <V> V lockObjectsForModification(KeycloakSession session, CallableWithoutThrowingAnException<V> callable) {
-        if (LockObjectsForModification.isEnabled(session)) {
+    private static <V> V lockObjectsForModification(KeycloakSession session, Class<?> model, CallableWithoutThrowingAnException<V> callable) {
```

Comment:
Whats the idea behind making this private? If we need to lock other areas later (e.g. auth sessions) we would add an area-specific method here, like `lockAuthSessionsForModification(session, callable)`?

Reply:
yes, that was the intention - add area specific methods as needed. 

I don't want people to pick the classes manually; they won't be type safe so they could place any class there. There is no base interface/class for the models AFAIK.